### PR TITLE
Adds unit testing and fixes supplier/borrower counts

### PR DESCRIPTION
**Pull request by lyoungblood**
This pull request does the following:
- Adds `matchstick-as` as a dev dependency for unit testing
- Adds unit tests for the following event types: (be sure to expand the diff for `comptroller.test.ts` as it is too large by default)
  - MarketListed
  - Mint
  - Transfer
  - Redeem
  - Borrow
  - RepayBorrow
- Adds `Makefile` support for local testing (`make test`) and local deployment (`make local`)
- Adds `Makefile` support for dev deployment (`make dev network=X`)
  - Be sure to see the `README.md` for details on environment variables for dev deployment
- Fixes inaccurate market supplier counts (see comments inline below)
- Fixes inaccurate market borrower counts (see comments inline below)

Note that the code coverage report generated by matchstick doesn't appear accurate - I will troubleshoot further, but it is missing the following handlers that are definitely tested:
- handleMint
- handleRedeem
```
Generating coverage report 📝

Handlers for source 'CToken':
Handler 'handleMint' is not tested.
Handler 'handleRedeem' is not tested.
Handler 'handleBorrow' is tested.
Handler 'handleRepayBorrow' is tested.
Handler 'handleLiquidateBorrow' is not tested.
Handler 'handleAccrueInterest' is not tested.
Handler 'handleNewReserveFactor' is not tested.
Handler 'handleTransfer' is tested.
Handler 'handleNewMarketInterestRateModel' is not tested.
Test coverage: 33.3% (3/9 handlers).

Handlers for source 'Feed':
Handler 'handleAnswerUpdated' is not tested.
Test coverage: 0.0% (0/1 handlers).

Handlers for source 'Comptroller':
Handler 'handleMarketListed' is tested.
Handler 'handleMarketEntered' is not tested.
Handler 'handleMarketExited' is not tested.
Handler 'handleNewCloseFactor' is not tested.
Handler 'handleNewCollateralFactor' is not tested.
Handler 'handleNewLiquidationIncentive' is not tested.
Handler 'handleNewPriceOracle' is not tested.
Handler 'handleSupplyRewardSpeedUpdated' is not tested.
Handler 'handleBorrowRewardSpeedUpdated' is not tested.
Handler 'handleNewBorrowCap' is not tested.
Handler 'handleMTokenActionPaused' is not tested.
Test coverage: 9.1% (1/11 handlers).

Handlers for source 'Governor':
Handler 'handleProposalCanceled' is not tested.
Handler 'handleProposalCreated' is not tested.
Handler 'handleProposalCreated' is not tested.
Handler 'handleProposalExecuted' is not tested.
Handler 'handleProposalQueued' is not tested.
Handler 'handleVoteCast' is not tested.
Handler 'handleBreakGlassGuardianChanged' is not tested.
Handler 'handleProposalThresholdChanged' is not tested.
Handler 'handleQuorumVotesChanged' is not tested.
Handler 'handleVotingDelayedChanged' is not tested.
Handler 'handleVotingPeriodChanged' is not tested.
Test coverage: 0.0% (0/11 handlers).

Global test coverage: 12.5% (4/32 handlers).
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,4 +7,7 @@ generated/
 build/
 # generated from mustache
 subgraph.yaml
-config/config.ts
\ No newline at end of file
+config/config.ts
+# testing specific
```
Comment (lyoungblood) [inline .gitignore:11]: Exclude binary files which are downloaded automatically by matchstick when you test locally.

File: Makefile [hunk 2]
```diff
@@ -23,7 +30,28 @@ build:
...
-	yarn graph deploy moonwell-fi/moonwell-$(network) --access-token $(access_token) --node https://api.thegraph.com/deploy/
+	yarn graph deploy moonwell-fi/moonwell-$(network) --access-token $(graph_access_token) --node https://api.thegraph.com/deploy/
+
+deploy-dev:
+	yarn graph deploy $(github_dev_profile)/moonwell-$(network) --access-token $(graph_dev_access_token) --node https://api.thegraph.com/deploy/
+
+deploy-local:
+	yarn graph deploy $(github_dev_profile)/moonwell-$(network) --node http://localhost:8020 --ipfs http://localhost:5001
+
+deploy-satsuma:
```
Comment (lyoungblood) [inline Makefile:42]: Please ignore this for now - it's useful for testing as the sync time is several hours instead of several days

File: Makefile [hunk 3]
```diff
@@ -23,7 +30,28 @@ build:
...
 
 all:
-	$(MAKE) clean generate-config codegen deploy
\ No newline at end of file
+	$(MAKE) clean generate-config codegen deploy
+
+dev:
+	$(MAKE) clean generate-config codegen deploy-dev deploy-satsuma
+
+local:
```
Comment (lyoungblood) [inline Makefile:54]: Requires you to have a local Graph node running in Docker  (or otherwise)

File: README.md [hunk 4]
```diff
@@ -21,3 +21,10 @@ You can also see the saved queries on the hosted service for examples.
 ## Development
 
 The subgraph is under major refactor in `dev` branch.
+
+## Local Dev Environment Setup
+
+Set the following environment variables:
```
Comment (lyoungblood) [inline README.md:27]: In order to test on a development subgraph, you should first create one on your personal Github profile (not Moonwell-fi) and set the environment variables as follows. Once you've set `GITHUB_DEV_PROFILE` to your Github username and set the access tokens, `make dev network=moonriver` or similar will deploy to both the hosted subgraph service and Satsuma.

File: package.json [hunk 5]
```diff
@@ -9,7 +9,8 @@
   },
   "devDependencies": {
     "@graphprotocol/graph-cli": "0.30.1",
-    "prettier": "^2.6.2",
-    "mustache": "^4.2.0"
+    "matchstick-as": "^0.5.2",
```
Comment (lyoungblood) [inline package.json:12]: We only add `matchstick-as` as a dev dependency.

File: src/comptroller.ts [hunk 6]
```diff
@@ -59,7 +59,6 @@ export function handleMarketEntered(event: MarketEntered): void {
     return
   }
 
-  market.borrowerCount = market.borrowerCount + 1
```
Comment (lyoungblood) [inline src/comptroller.ts:62]: Previous logic used the `MarketEntered` event handler to increment the `borrowerCount` in the market, which is inaccurate. Entering a market only allows your balance in cTokens to be used for borrowing power calculations, and has no correlation with whether a given account has borrowed from that market or not.

File: src/comptroller.ts [hunk 7]
```diff
@@ -89,7 +88,6 @@ export function handleMarketExited(event: MarketExited): void {
     return
   }
 
-  market.borrowerCount = market.borrowerCount - 1
```
Comment (lyoungblood) [inline src/comptroller.ts:92]: Previous logic used the `MarketExited` event handler to decrement the `borrowerCount` in the market, which is inaccurate. Exiting a market only removes your balance in cTokens from being used for borrowing power calculations, and has no correlation with whether a given account has borrowed from that market or not.

File: src/ctoken.ts [hunk 8]
```diff
@@ -29,6 +29,8 @@ import {
   exponentToBigDecimal,
   cTokenDecimalsBD,
   cTokenDecimals,
+  zeroBI,
```
Comment (lyoungblood) [inline src/ctoken.ts:32]: These are just some helpers that make it easier to compare against a BigInt zero or a BigDecimal zero

File: src/ctoken.ts [hunk 9]
```diff
@@ -157,6 +159,17 @@ export function handleBorrow(event: Borrow): void {
     cTokenStats.totalUnderlyingBorrowed.plus(borrowAmountBD)
   cTokenStats.save()
 
+  if (cTokenStats.totalUnderlyingBorrowed.equals(borrowAmountBD)) {
```
Comment (lyoungblood) [inline src/ctoken.ts]: If the `borrowAmount` in the `Borrow` event is equal to the `totalUnderlyingBorrowed` by that account, it indicates that before this borrow the `totalUnderlyingBorrowed` was 0. If that is the case, this is a new borrower so we should increment `borrowerCount`

File: src/ctoken.ts [hunk 10]
```diff
@@ -157,6 +159,17 @@ export function handleBorrow(event: Borrow): void {
...
+    // If this is the first borrow, increment the market's borrower count
+    log.warning('[handleBorrow] +1 cTokenStats.totalUnderlyingBorrowed.equals(borrowAmountBD) marketID: {}, borrower: {}, amount: {}', [
```
Comment (lyoungblood) [inline src/ctoken.ts]: I've added `log.warning` calls whenever we increment/decrement `borrowerCount` so that we can analyze all the events for data integrity. We might remove these when we have greater confidence in our data integrity, or if it harms indexing performance (not sure tbh).

File: src/ctoken.ts [hunk 11]
```diff
@@ -204,6 +217,24 @@ export function handleRepayBorrow(event: RepayBorrow): void {
     createAccount(accountID)
   }
 
+  if (
+    event.params.accountBorrows
+      .toBigDecimal()
+      .div(exponentToBigDecimal(market.underlyingDecimals))
+      .truncate(market.underlyingDecimals)
+      .equals(zeroBD)
+  ) {
```
Comment (lyoungblood) [inline src/ctoken.ts]: `accountBorrows` is an event parameter set by the market contract that is set to the stored borrow balance after the borrow is repaid. If it is equal to zero that indicates this borrower no longer has any active borrow in this market, so we should decrement the `borrowerCount`.

File: src/ctoken.ts [hunk 12]
```diff
@@ -360,6 +391,12 @@ export function handleTransfer(event: Transfer): void {
     ])
   } else {
     if (balanceOfToAccountResult.value.equals(event.params.amount)) {
```
Comment (lyoungblood) [inline src/ctoken.ts:387]: When handling `Transfer` events, if the resulting balance of the `to` account is the same as the `amount` of the transfer, that indicates the starting balance of the `to` account was 0, so they are technically a new supplier to that market (by virtue of simply holding the cTokens). If true, this will increment `supplierCount`, irrespective of the balance of the `source` account.

File: src/ctoken.ts [hunk 13]
```diff
@@ -372,7 +409,15 @@ export function handleTransfer(event: Transfer): void {
       marketID,
     ])
   } else {
-    if (balanceOfFromAccountResult.value.equals(event.params.amount)) {
+    // We should only decrement supplierCount when the from balance is now zero -AND- the from is not the market contract
+    // We should not decrement supplier count when the from is the market contract (i.e. minting)
+    if ((balanceOfFromAccountResult.value.equals(zeroBI) && !event.params.from.equals(cTokenContract._address))) {
```
Comment (lyoungblood) [inline src/ctoken.ts:408]: When handling `Transfer` events, if the resulting balance of the `from` account is now 0 **AND** the `from` account is not the market contract (market contracts will also have a cToken balance of 0 unless someone has inadvertently sent some cTokens to it) we should decrement the `supplierCount` for that market.

The bug that caused negative supplier counts was due to the fact that `mint` calls initiate a `transfer` from the market contract to the minter. Because the `USDC.multi` market has a 0 `balanceOf` for its own cToken, this would cause mints to incorrectly decrement the `supplierCount`.

There is a unit test that mocks a cToken transfer from test account 1 to test account 2 and mocks the `balanceOf` contract calls. This type of transfer will now correctly decrement `supplierCount` **AND** increment `supplierCount` as both of these conditions are triggered on the `handleTransfer`.

Since all types of events that can result in cToken movement emit a `Transfer`, this should encompass all edge cases.

File: tests/comptroller/comptroller.test.ts [hunk 14]
```diff
@@ -0,0 +1,779 @@
...
+    handleNewTransfers([newTransferEvent])
+    assert.entityCount("TransferEvent", 1) // There should be one transfer now
+    assert.fieldEquals("Market", movrMarketAddress, "supplierCount", "1") // There should be one supplier now
+    assert.fieldEquals("TransferEvent", transferID, "id", transferID)
+    assert.fieldEquals("TransferEvent", transferID, "from", movrMarketAddress)
+    assert.fieldEquals("TransferEvent", transferID, "to", testAccountAddress1)
+    assert.fieldEquals("TransferEvent", transferID, "amount", transferAmount.toBigDecimal().div(cTokenDecimalsBD).toString())
+  })
+
+  test("Handle a Liquidation-type 1 (seize mTokens) Transfer event (liquidated -> liquidator)", () => {
```
Comment (lyoungblood) [inline tests/comptroller/comptroller.test.ts:504]: This is the test case that will both decrement `supplierCount` and increment `supplierCount` as the cTokens are transferred between two EOA accounts. (see above)

File: tests/comptroller/utils.ts [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+  let newMint = changetype<Mint>(newMockEventWithParams([minterParam, mintAmountParam, mintTokensParam]));
+  newMint.address = Address.fromString(to); // Set the address of the event to the mToken address
+  newMint.transaction.hash = Bytes.fromHexString(hash) as Bytes; // Set the hash of the transaction
+  return newMint;
+}
+
+export function handleNewMints(events: Mint[]): void {
+  events.forEach(event => {
+    handleMint(event)
```
Comment (lyoungblood) [inline tests/comptroller/utils.ts:81]: We're definitely covering `handleMint` - not sure what is up with code coverage but I'll troubleshoot more.

File: tests/comptroller/utils.ts [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+  let newRedeem = changetype<Redeem>(newMockEventWithParams([redeemerParam, redeemAmountParam, redeemTokensParam]));
+  newRedeem.address = Address.fromString(to); // Set the address of the event to the mToken address
+  newRedeem.transaction.hash = Bytes.fromHexString(hash) as Bytes; // Set the hash of the transaction
+  return newRedeem;
+}
+
+export function handleNewRedeems(events: Redeem[]): void {
+  events.forEach(event => {
+    handleRedeem(event)
```
Comment (lyoungblood) [inline tests/comptroller/utils.ts:115]: We're definitely covering `handleRedeem` - not sure what is up with code coverage but I'll troubleshoot more.

**Review by lyoungblood — COMMENTED** [review]
Here are my review comments. Please let me know if you still have questions about this after reviewing.

File: src/ctoken.ts [hunk 17]
```diff
@@ -157,6 +159,17 @@ export function handleBorrow(event: Borrow): void {
     cTokenStats.totalUnderlyingBorrowed.plus(borrowAmountBD)
   cTokenStats.save()
 
+  if (event.params.accountBorrows.equals(event.params.borrowAmount)) {
+    // If this is the first borrow, increment the market's borrower count
```
Comment (lyoungblood) [inline src/ctoken.ts:163]: Determine whether this is the first borrow if the resulting `accountBorrows` param is equal to the `borrowAmount`. This should only increment if the borrower has no existing borrows in the market.

File: src/ctoken.ts [hunk 18]
```diff
@@ -204,6 +217,18 @@ export function handleRepayBorrow(event: RepayBorrow): void {
     createAccount(accountID)
   }
 
+  if (event.params.accountBorrows.equals(zeroBI)) {
```
Comment (lyoungblood) [inline src/ctoken.ts:220]: `accountBorrows` should always be a BigInt 0 if the loan is fully repaid, so we should decrement `market.borrowerCount`
